File: test/exp_logs.js
// ------------------------------------------------------------------------
// Copyright (c) 2019-2024 Alexandre Bento Freire. All rights reserved.
// Licensed under the MIT License
// ------------------------------------------------------------------------

/**
 * Mocha tests for numnjstoix Exponents and logarithms functions.
 */

// ------------------------------------------------------------------------
//                               Requires
// ------------------------------------------------------------------------

import { numnjstoix } from '../main.js';
import * as shared from './shared.js';

// ------------------------------------------------------------------------
//                               Test Cases
// ------------------------------------------------------------------------

describe('Exponents and logarithms', function () {

  shared.addTransformCase(describe, 'exp', Math.exp, numnjstoix.exp, true);
  shared.addTransformCase(describe, 'expm1', Math.expm1, numnjstoix.expm1, true);
  shared.addTransformCase(describe, 'exp2', (v) => Math.pow(2, v), numnjstoix.exp2, true);

  shared.addTransformCase(describe, 'log', Math.log, numnjstoix.log, true);
  shared.addTransformCase(describe, 'log10', Math.log10, numnjstoix.log10, true);
  shared.addTransformCase(describe, 'log2', Math.log2, numnjstoix.log2, true);
});

File: src/exp_logs.cpp
// ------------------------------------------------------------------------
// Copyright (c) 2019-2024 Alexandre Bento Freire. All rights reserved.
// Licensed under the MIT License
// ------------------------------------------------------------------------

// ------------------------------------------------------------------------
//                               Includes
// ------------------------------------------------------------------------

#include "numnjs.h"
#include <math.h>
#include <node.h>

namespace numnjs {

// ------------------------------------------------------------------------
//                               Actual Functions
// ------------------------------------------------------------------------

/* clang-format off */
transfCalc1DblFunc(expFunc, exp)
transfCalc1DblFunc(expm1Func, expm1)
transfCalc1DblFunc(exp2Func, exp2)

transfCalc1DblFunc(logFunc, log)
transfCalc1DblFunc(log10Func, log10)
transfCalc1DblFunc(log2Func, log2)
/* clang-format on */

// ------------------------------------------------------------------------
//                               Initialize
// ------------------------------------------------------------------------

void initializeExpLogs(v8exports exports) {

  NODE_SET_METHOD(exports, "exp", expFunc);
  NODE_SET_METHOD(exports, "expm1", expm1Func);
  NODE_SET_METHOD(exports, "exp2", exp2Func);

  NODE_SET_METHOD(exports, "log", logFunc);
  NODE_SET_METHOD(exports, "log10", log10Func);
  NODE_SET_METHOD(exports, "log2", log2Func);
}

} // namespace numnjs